src/Rings/MPolyMap/AffineAlgebras.jl

```diff
@@ -0,0 +1,220 @@
+
+affine_algebra_morphism_type(R::S, U::T) where {S, T} = affine_algebra_morphism_type(S, T)
+
+################################################################################
+#
+#  Singular data stuff
+#
+################################################################################
+
+function _singular_ring_domain(f::MPolyAnyMap)
+  return get_attribute!(f, :singular_ring_domain) do
+    singular_ring(domain(f))
+  end
+end
+
+function _singular_ring_codomain(f::MPolyAnyMap)
+  return get_attribute!(f, :singular_ring_codomain) do
+    singular_ring(codomain(f))
+  end
+end
+
+function _singular_algebra_morphism(f::MPolyAnyMap)
+  return get_attribute!(f, :singular_algebra_morphism) do
+    DS = _singular_ring_domain(f)
+    CS = _singular_ring_codomain(f)
+    CSimgs = CS.(_images(f))
+    return Singular.AlgebraHomomorphism(DS, CS, CSimgs)
+  end
+end
+
+################################################################################
+#
+#  Preimage of ideal
+#
+################################################################################
+
+function preimage(f::AffAlgHom, I::Union{MPolyIdeal, MPolyQuoIdeal})
+  @req base_ring(I) === codomain(f) "Parent mismatch"
+  D = domain(f)
+  salghom = _singular_algebra_morphism(f)
+  CS = codomain(salghom)
+  V = gens(I)
+  Ix = Singular.Ideal(CS, CS.(V))
+  prIx = Singular.preimage(salghom, Ix)
+  return ideal(D, D.(gens(prIx)))
+end
+
+################################################################################
+#
+#  Kernel
+#
+################################################################################
+
+function kernel(f::AffAlgHom)
+  get_attribute!(f, :kernel) do
+    C = codomain(f)
+    return preimage(f, ideal(C, [zero(C)]))
+  end # TODO: need some ideal_type(domain(f)) here :)
+end
+
+##############################################################################
+#
+#  Injectivity
+#
+##############################################################################
+
+@doc Markdown.doc"""
+    isinjective(F::AlgHom)
+
+Return `true` if `F` is injective, `false` otherwise.
+"""
+function isinjective(F::AffAlgHom)
+   G = gens(kernel(F))
+   for i in 1:length(G)
+      !iszero(G[i]) && return false
+   end
+   return true
+end
+
+# Helper function related to the computation of surjectivity, preimage etc.
+# Stores the necessary data in groebner_data, resp. groebner_data_lex
+function _groebner_data(F::AffAlgHom, ord::Symbol)
+   r = domain(F)
+   s = codomain(F)
+   n = ngens(r)
+   m = ngens(s)
+   return get_attribute!(F, ord) do
+     (S, I, W, _) = _ring_helper(s, zero(s), _images(F))
+     # Build auxilliary objects
+     (T, inc, J) = _containement_helper(S, n, m, I, W, ord)
+     D = normal_form([gen(T, i) for i in 1:m], J)
+     A = [zero(r) for i in 1:m]
+     B = [gen(r, i) for i in 1:n]
+     pr = hom(T, r, vcat(A, B))
+     groebner_data_lex = (T, inc, pr, J, D)
+     return groebner_data_lex
```

Comment:
Is the name "groebner_data_lex" just a leftover from an older revision of this patch, or is there a deeper meaning that I am missing? It seems this code is used now also for degrevlex and arbitrary ordering?!?

For that matter...
```suggestion
     return (T, inc, pr, J, D)
```